login.php

```diff
@@ -29,7 +29,7 @@
 $responsetype = required_param('response_type', PARAM_RAW);
 $scope = optional_param('scope', false, PARAM_TEXT);
 $state = optional_param('state', false, PARAM_TEXT);
-$redirect_uri = urlencode(optional_param('redirect_uri', false, PARAM_URL));
+$redirect_uri = urlencode(optional_param('redirect_uri', false, PARAM_RAW));
```

Comment:
Will need to change to `rawurlencode` if `redirect_uri` spaces should be `%20` instead of `+`

```suggestion
$redirect_uri = rawurlencode(optional_param('redirect_uri', false, PARAM_RAW));
```